runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/SplittableRemoteStageEvaluatorFactory.java

```diff
@@ -0,0 +1,179 @@
+
+  // A fictional transform that transforms from KWI<unique key, KV<element, restriction>>
+  // to simply KV<element, restriction> taken by the SDF inside the ExecutableStage.
+  public static final String FEED_SDF_URN = "urn:beam:directrunner:transforms:feed_sdf:v1";
+
+  private final BundleFactory bundleFactory;
+  private final JobBundleFactory jobBundleFactory;
+  private final StepStateAndTimers.Provider stp;
+
+  SplittableRemoteStageEvaluatorFactory(
+      BundleFactory bundleFactory,
+      JobBundleFactory jobBundleFactory,
+      StepStateAndTimers.Provider stepStateAndTimers) {
+    this.bundleFactory = bundleFactory;
+    this.jobBundleFactory = jobBundleFactory;
+    this.stp = stepStateAndTimers;
+  }
+
+  @Nullable
+  @Override
+  public <InputT> TransformEvaluator<InputT> forApplication(
+      PTransformNode application, CommittedBundle<?> inputBundle) throws Exception {
+    return new SplittableRemoteStageEvaluator(
+        bundleFactory,
+        jobBundleFactory,
+        stp.forStepAndKey(application, inputBundle.getKey()),
+        application);
+  }
+
+  @Override
+  public void cleanup() throws Exception {
+    jobBundleFactory.close();
+  }
+
+  private static class SplittableRemoteStageEvaluator<InputT, RestrictionT>
+      implements TransformEvaluator<KeyedWorkItem<byte[], KV<InputT, RestrictionT>>> {
+    private final PTransformNode transform;
+    private final ExecutableStage stage;
+
+    private final CopyOnAccessInMemoryStateInternals<byte[]> stateInternals;
+    private final DirectTimerInternals timerInternals;
+    private final RemoteBundle<KV<InputT, RestrictionT>> bundle;
+    private final Collection<UncommittedBundle<?>> outputs;
+
+    private final SDFFeederViaStateAndTimers<InputT, RestrictionT> feeder;
+
+    private SplittableRemoteStageEvaluator(
+        BundleFactory bundleFactory,
+        JobBundleFactory jobBundleFactory,
+        StepStateAndTimers<byte[]> stp,
+        PTransformNode transform)
+        throws Exception {
+      this.stateInternals = stp.stateInternals();
+      this.timerInternals = stp.timerInternals();
+      this.transform = transform;
+      this.stage =
+          ExecutableStage.fromPayload(
+              ExecutableStagePayload.parseFrom(transform.getTransform().getSpec().getPayload()));
+      this.outputs = new ArrayList<>();
+      this.bundle =
+          jobBundleFactory
+              .<KV<InputT, RestrictionT>>forStage(stage)
+              .getBundle(
+                  BundleFactoryOutputReceiverFactory.create(
+                      bundleFactory, stage.getComponents(), outputs::add),
+                  StateRequestHandler.unsupported(),
+                  new BundleProgressHandler() {
+                    @Override
+                    public void onProgress(ProcessBundleProgressResponse progress) {
+                      if (progress.hasSplit()) {
+                        feeder.split(progress.getSplit());
```

Comment:
nit: it would make me happier if this.feeder was initialized before the RemoteBundle is created so we don't have to worry about NPE.